### ISSUE TITLE
[prefer-const] false positive when casting left side of assignment operator

### ISSUE THREAD

**Issue by IlyaSemenov** [typescript-eslint/typescript-eslint#1615]
**Repro**

```JS
module.exports = {
	root: true,
	extends: ['plugin:@typescript-eslint/recommended'],
}
```

```TS
let foo = 1
;(foo as number) = 2
```

**Expected Result**

No errors.

**Actual Result**

'foo' is never reassigned. Use 'const' instead. eslint(prefer-const)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.20.0` |
| `@typescript-eslint/parser`        | `2.20.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.9.0` |

**Comment by IlyaSemenov** [comment]
I understand it's a rule that comes from eslint (not typescript-eslint) however since it has to do with Typescript syntax I believe one could expect typescript-eslint to add workarounds and support this kind of syntax (especially in recommended config).

**Comment by bradzacher** [comment]
This seems like a pretty big code smell.
How come you're casting the left hand side of an assignment expression?

**Comment by IlyaSemenov** [comment]
Honestly I don't understand what it has to do with the problem. Even though this is 100% code smell (I already yelled at my colleagues over it, lol), the current implementation **does** break the working code when run with `eslint --fix`. Regardless of how ugly the code is, it should not become broken by the linter... shouldn't it?

**Comment by bradzacher** [comment]
There's no such thing as "just fixing" this, it involves creating an extension rule in this plugin, and figuring out how to hack the existing rule implementation to support it.
Sometimes this is easy, sometimes not.

Having a quick look at the source code for `prefer-const`, it looks like it's not a simple hack fix like many of our other extension rules. This rule relies upon scope analysis, which we do not support well.
So fixing this case is more than likely going to be hours of work.

Given that:
- this is a code smell, and should never happen in code.
- I've never seen anyone do this before in my ~5 years of coding in TS.
- I'll probably never see anyone do this again.

It's probably not something that's worth that much effort.

**Comment by IlyaSemenov** [comment]
I totally understand this reasoning. I don't demand anything, after all. I support a number of open source libraries myself, so I pretty much understand how things work.

This is why there is a backlog of problems, ranked by priority. This is 100% not top priority and I realise that depending on how much people are involved this may never be fixed.

What I don't understand is when/if edge cases like this are disregarded. After all, I deliberately trimmed it down to the absolute minimum to demonstrate the problem... and now you say it doesn't make sense to do it this way. Of course it doesn't, most code doesn't make sense when trimmed down. It made a bit more sense in the full context (although right-hand casting of course better).

**Comment by bradzacher** [comment]
We currently have a backlog of 206 issues including this one.
This backlog grows every day.
This project is run by volunteer maintainers, and lives off the back of volunteer contributors.

I'm never going to work on this, I doubt another maintainer is going to pick it up, and I doubt a random contributor is going to pick it up - all for the same reasons of this being a lot of work to fix, and a very, very, very rare edge.

Which is why on many issues I push back on bad code patterns; it's often easier to make someone write proper code that doesn't break than it is to bend the tooling to handle the bad code pattern.
It's the tradeoff we have from existing in a mixed ecosystem.

Happy to leave this open, but as mentioned, unless you're willing to put up a fix, I doubt anyone will ever work on it.

**Comment by IlyaSemenov** [comment]
Thank you, I couldn't demand more. Who knows, may be one day I will actually learn the internals of eslint, just for the sake of it ;-)

**Comment by runspired** [comment]
After upgrading typescript-eslint parser and plugin:

```
-    "@typescript-eslint/eslint-plugin": "^2.32.0",
-    "@typescript-eslint/parser": "^2.33.0",
+    "@typescript-eslint/eslint-plugin": "^4.21.0",
+    "@typescript-eslint/parser": "^4.21.0",
```

I encounter this in situations that did not error before.

For instance the following will now trigger this lint rule

```js
let lazilyDefinedCallback:  (...args: any[]) => void;

function usesTheCallback() {
   lazilyDefinedCallback();
}

// some other code in the same scope that the callback will make use of

lazilyDefinedCallback = () => {
   // my stuff
};
```

**Comment by Josh-Cena** [comment]
This no longer seems to report on the latest ESLint. I wonder why, but closing as completed.
